### ISSUE TITLE
Apply pmap to a list of pytrees

### ISSUE THREAD

**Issue by awav** [google/jax#2540]
Hello all!

I can't make `pmap` work with `pytrees` objects. Although, `jit` and `grad` transformations work fine with it. I found https://github.com/google/jax/issues/1486 and https://github.com/google/jax/pull/1514, but that does't help much with solving the issue.

[Here is a simple MWE](https://colab.research.google.com/drive/1uSsZs9PBEhT6IuSLyTCMbyM7phGso9w9).

Thanks!

**Comment by mattjj** [comment]
Thanks for the question!

The problem is that `v` isn't a _list_ of pytrees, but rather a numpy-ndarray-of-dtype-object of pytrees. None of the JAX API functions will work with an object like that, but all of them will work with a list of pytrees.

In your notebook, you can just change this:

```python
v = onp.array([TrainableParameter(2.0), TrainableParameter(3.0)])
```

to this:

```python
v = [TrainableParameter(2.0), TrainableParameter(3.0)]
```

and you won't see the same error.

You will, however, see two other errors!
1. `pmap` needs a leading axis to map over
2. the function `jax.pmap(lambda x: jnp.exp(x))` has a type error when applied to inputs that are a list of `TrainableParameter`s, so we need to unpack the list and the `TrainableParameter` before applying `jnp.exp` (which only applies to arrays)

You can fix the first like:

```python
v = [TrainableParameter(jnp.array([2.0])), TrainableParameter(jnp.array([3.0]))]
```

and the second like:

```python
jax.pmap(lambda lst: jnp.exp(lst[0]()))(v)
```

On colab, unless you use our [Cloud TPU preview](https://github.com/google/jax/tree/master/cloud_tpu_colabs), you'll only have access to one XLA device. You can also use it with a multi-GPU machine, e.g. from a cloud provider.

WDYT?

**Comment by mattjj** [comment]
I'm going to close this issue because I think we answered the question from the colab, and also because we currently have 666 closed issues and I can't let it sit at that number. Please open a new issue if more questions arise!

**Comment by ThatBitBoss** [comment]
[I am strugling](https://colab.research.google.com/drive/19W8lWc5QcDajag5Jkj8-PQ1Ph9N98gH6#scrollTo=rLiHnhbgvjcR&line=1&uniqifier=1)

I am trying to pmap a list of pytree and keep getting an exception. The goal would be to map over execGoal_HUNGER but even a simple lambda with print seem to fail. Can you help?

**Comment by jakevdp** [comment]
@ThatBitBoss answered in #19133
